lib/assertions.js

```diff
@@ -1150,60 +1150,74 @@ module.exports = expect => {
 
       const subjectType = expect.findTypeOf(subject);
       const keys = subjectType.getKeys(subject);
-      return expect.promise
-        .any(
-          keys.map((key, index) => {
-            let expected;
-            if (typeof nextArg === 'string') {
-              expected = s => expect.shift(s);
-            } else if (typeof nextArg === 'function') {
-              expected = s => nextArg(s, index);
-            } else {
-              expected = nextArg;
-            }
-            return expect.promise(() =>
-              expect(
-                subjectType.valueForKey(subject, key),
-                'to [exhaustively] satisfy',
-                expected
-              )
-            );
-          })
-        )
-        .caught(err =>
-          expect.fail(output => {
-            output.append(
-              expect.standardErrorMessage(output.clone(), {
-                compact: err && err._isUnexpected && err.hasDiff()
+      const not = !!expect.flags.not;
+
+      const keyResults = new Array(keys.length);
+
+      expect.withError(
+        () =>
+          expect.promise[not ? 'all' : 'any'](
+            keys.map((key, index) => {
+              let expected;
+              if (typeof nextArg === 'string') {
+                expected = s => expect.shift(s);
+              } else if (typeof nextArg === 'function') {
+                expected = s => nextArg(s, index);
+              } else {
+                expected = nextArg;
+              }
+
+              keyResults[key] = expect.promise(() =>
+                expect(
+                  subjectType.valueForKey(subject, key),
+                  '[not] to [exhaustively] satisfy',
+                  expected
+                )
+              );
+              return keyResults[key];
+            })
+          ),
+        err => {
+          expect.fail({
+            message(output) {
+              output.append(
+                expect.standardErrorMessage(output.clone(), {
+                  compact: err && err._isUnexpected && err.hasDiff()
+                })
+              );
+            },
+            diff:
+              expect.flags.not &&
+              ((output, diff, inspect, equal) => {
+                const expectedObject = subjectType.name === 'array' ? [] : {};
```

Comment:
Is there any smarter way to do this? Ideally it should be array-like and friends. Can i somehow query the type to see if it is a subtype of array-like?

Reply:
Yes: `subjectType.is('array-like')`